File: lib_base/src/main/java/com/android/base/adapter/recycler/MultiTypeAdapter.java
package com.android.base.adapter.recycler;

import android.content.Context;

import com.android.base.foundation.adapter.DataManager;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


/**
 * @see <a href='https://github.com/drakeet/MultiType'>drakeet/MultiTypeAdapter</a>
 */
public class MultiTypeAdapter extends com.drakeet.multitype.MultiTypeAdapter implements DataManager<Object> {

    private final Context mContext;

    private RecyclerDataManagerImpl<Object> mRecyclerDataManager;

    public MultiTypeAdapter(Context context) {
        super();
        mContext = context;
        ArrayList<Object> objects = new ArrayList<>();
        mRecyclerDataManager = new RecyclerDataManagerImpl<>(objects, this);
        super.setItems(objects);
    }

    public MultiTypeAdapter(Context context, @NonNull List<?> items) {
        super();
        mContext = context;
        ArrayList<Object> objects = new ArrayList<>(items);
        mRecyclerDataManager = new RecyclerDataManagerImpl<>(objects, this);
        super.setItems(objects);
    }

    public MultiTypeAdapter(Context context, @NonNull List<?> items, int initialCapacity) {
        super(items, initialCapacity);
        mContext = context;
        ArrayList<Object> objects = new ArrayList<>(items);
        mRecyclerDataManager = new RecyclerDataManagerImpl<>(objects, this);
        super.setItems(objects);
    }

    @SuppressWarnings("unused")
    public void notifyEntryChanged(Object entry) {
        int itemPosition = indexItem(entry);
        if (itemPosition != -1) {
            notifyItemChanged(itemPosition);
        }
    }

    @Override
    public void add(Object elem) {
        mRecyclerDataManager.add(elem);
    }

    @Override
    public void addAt(int location, Object elem) {
        mRecyclerDataManager.addAt(location, elem);
    }

    @Override
    public void addItems(List<Object> elements) {
        mRecyclerDataManager.addItems(elements);
    }

    @Override
    public void addItemsChecked(List<Object> elements) {
        mRecyclerDataManager.addItemsChecked(elements);
    }

    @Override
    public void addItemsAt(int location, List<Object> elements) {
        mRecyclerDataManager.addItemsAt(location, elements);
    }

    @Override
    public void replace(Object oldElem, Object newElem) {
        mRecyclerDataManager.replace(oldElem, newElem);
    }

    @Override
    public void replaceAt(int index, Object elem) {
        mRecyclerDataManager.replaceAt(index, elem);
    }

    @Override
    public void replaceAll(List<Object> elements) {
        mRecyclerDataManager.replaceAll(elements);
    }

    @Override
    public void setDataSource(@NonNull List<Object> newDataSource, boolean notifyDataSetChanged) {
        super.setItems(newDataSource);
        mRecyclerDataManager.setDataSource(newDataSource, notifyDataSetChanged);
    }

    @Override
    public void remove(Object elem) {
        mRecyclerDataManager.remove(elem);
    }

    @Override
    public void removeAt(int index) {
        mRecyclerDataManager.removeAt(index);
    }

    @Override
    public void removeItems(List<Object> elements) {
        mRecyclerDataManager.removeItems(elements);
    }

    @Override
    public void removeItems(List<Object> elements, boolean isSuccessive) {
        mRecyclerDataManager.removeItems(elements, isSuccessive);
    }

    @Nullable
    @Override
    public Object getItem(int position) {
        return mRecyclerDataManager.getItem(position);
    }

    @Override
    public int getDataSize() {
        return mRecyclerDataManager.getDataSize();
    }

    @Override
    public boolean contains(Object elem) {
        return mRecyclerDataManager.contains(elem);
    }

    @Override
    public boolean isEmpty() {
        return mRecyclerDataManager.isEmpty();
    }

    @Override
    public void clear() {
        mRecyclerDataManager.clear();
    }

    @NonNull
    @Override
    public List<Object> getItems() {
        return mRecyclerDataManager.getItems();
    }

    @Override
    public int indexItem(Object o) {
        return mRecyclerDataManager.indexItem(o);
    }

    @Override
    public void setItems(@NonNull List<?> items) {
        ArrayList<Object> objects = new ArrayList<>(items);
        super.setItems(objects);
        mRecyclerDataManager.setDataSource(objects, true);
    }

    public Context getContext() {
        return mContext;
    }

}

File: lib_base/src/main/java/com/android/base/adapter/recycler/MultiTypeAdapter.kt
package com.android.base.adapter.recycler

import android.content.Context
import com.android.base.foundation.adapter.DataManager

/**
 * @see [MultiTypeAdapter](https://github.com/drakeet/MultiType)
 */
open class MultiTypeAdapter : com.drakeet.multitype.MultiTypeAdapter, DataManager<Any> {

    val context: Context

    private var dataManager: RecyclerDataManagerImpl<Any>

    @Suppress("LeakingThis")
    constructor(context: Context) : super() {
        this.context = context
        val items = mutableListOf<Any>()
        dataManager = RecyclerDataManagerImpl(items, this)
        super.items = items
    }

    @Suppress("LeakingThis")
    constructor(context: Context, items: List<Any>) : super() {
        this.context = context
        val mutableList = items.toMutableList()
        dataManager = RecyclerDataManagerImpl(mutableList, this)
        super.items = mutableList
    }

    override fun add(element: Any) {
        dataManager.add(element)
    }

    override fun addAt(location: Int, element: Any) {
        dataManager.addAt(location, element)
    }

    override fun addItems(elements: List<Any>) {
        dataManager.addItems(elements)
    }

    override fun addItemsChecked(elements: List<Any>) {
        dataManager.addItemsChecked(elements)
    }

    override fun addItemsAt(location: Int, elements: List<Any>) {
        dataManager.addItemsAt(location, elements)
    }

    override fun replace(oldElement: Any, newElement: Any) {
        dataManager.replace(oldElement, newElement)
    }

    override fun replaceAt(index: Int, element: Any) {
        dataManager.replaceAt(index, element)
    }

    override fun replaceAll(elements: List<Any>) {
        dataManager.replaceAll(elements)
    }

    override fun setDataSource(newDataSource: MutableList<Any>) {
        super.items = newDataSource
        dataManager.setDataSource(newDataSource)
    }

    override fun remove(element: Any): Boolean {
        return dataManager.remove(element)
    }

    override fun removeAt(index: Int) {
        dataManager.removeAt(index)
    }

    override fun removeItems(elements: List<Any>) {
        dataManager.removeItems(elements)
    }

    override fun removeItems(elements: List<Any>, isSuccessive: Boolean) {
        dataManager.removeItems(elements, isSuccessive)
    }

    override fun getItem(position: Int): Any? {
        return dataManager.getItem(position)
    }

    override operator fun contains(element: Any): Boolean {
        return dataManager.contains(element)
    }

    override fun clear() {
        dataManager.clear()
    }

    override fun indexItem(element: Any): Int {
        return dataManager.indexItem(element)
    }

    override fun notifyElementChanged(element: Any) {
        dataManager.notifyElementChanged(element)
    }

    override fun isEmpty(): Boolean {
        return dataManager.isEmpty()
    }

    override fun getList(): List<Any> {
        return dataManager.getList()
    }

    override fun getDataSize(): Int {
        return dataManager.getDataSize()
    }

    override var items: List<Any>
        get() = super.items
        set(value) {
            if (value is MutableList) {
                super.items = value
                dataManager.setDataSource(value)
            } else {
                throw IllegalArgumentException("only support MutableList")
            }
        }

    override fun removeIf(filter: (Any) -> Boolean) {
        dataManager.removeIf(filter)
    }

    override fun swipePosition(fromPosition: Int, toPosition: Int) {
        dataManager.swipePosition(fromPosition, toPosition)
    }

}